### DISCUSSION TITLE
GPU detection hangs when starting BOINC

### DISCUSSION THREAD

**Discussion by etdr** [BOINC/boinc#5187] [Q&A] [answered]
**Describe the bug**
I am inferring the GPU detection is the problem because of how the logs are split relative to when I shut the service down. (See screenshots.)

I might be missing something really obvious, or it might have to do with the systemd unit, but wanted to make sure no one knew why this was happening here.

**Steps To Reproduce**
1. Install the `boinc` package from Arch repository
2. Enable and start the `boinc-client.service` unit
3. The system journal never gets past the first picture below.

**Expected behavior**
Even if GPU detection fails, I expect it to come back with a yes or a no, not to hang.

**Screenshots**
![this is what it shows up until I kill it](https://user-images.githubusercontent.com/3358909/230511107-f294a6bb-8de5-4ced-9389-3e8f74d5476f.png)

![this is what it shows after I kill the service](https://user-images.githubusercontent.com/3358909/230511175-337b8c6d-16de-465f-9d13-7a5e9b2d324c.png)

**System Information**
 - OS: Arch Linux with Linux 6.2.9
 - BOINC Version: 7.22

**Additional context**
I have two GPUs in this machine, one AMD and one NVIDIA, both pretty recent. I have the most up-to-date libraries for opencl installed for both of them.

Nothing shows up in the stderrgpudetect file.

**Comment by davidpanderson** [comment]
If you run
boinc --gpu_detect
it will do GPU detection, write the results to coproc_info.xml, and exit.
I'd like to know if this crashes (and if so, where).

But the client is supposed to keep going even if the GPU detection subprocess fails.
Instead, it's being sent a SIGTERM (signal 14) and this causes it to exit.
We need to figure out who's sending this signal.

**Comment by etdr** [comment]
The SIGTERM is because I'm exiting it because it's hanging, because I didn't think GPU detection should take >5 minutes, I should have been clearer about that.

This is what I get when I try the `--gpu_detect` option:
![image](https://user-images.githubusercontent.com/3358909/230652263-7cf733ae-610f-42e4-b493-20cb7ab90541.png)

There is a `--no_gpus` option, though, but that is for explicitly not checking for GPUs?

**Comment by BrianNixon** [comment]
Try `--detect_gpus` instead of `--gpu_detect`

**Comment by etdr** [comment]
Just so I'm not exiting prematurely, how long should this take, maximum?

**Comment by BrianNixon** [comment]
“Not that long”, probably… (I don’t know the real answer)

But if the detection hangs, the client will also hang because it waits forever for the detection to complete. A timeout here would clearly be a good idea.

**Comment by etdr** [comment]
OK, it's just hanging then. Both running as boinc user and as root user. I never get that coproc_info.xml file that @davidpanderson was looking for.

I have added the boinc user to the video group, but maybe it's a permissions issue separate from that? Not sure.

**Comment by BrianNixon** [comment]
Not sure what else to suggest. You could try [the forum](https://boinc.berkeley.edu/forum_forum.php?id=10); maybe somebody there has some ideas.

During detection, warning messages are stored for reporting back to the client on completion. It would be useful if those also got written to the detector process’s `stderr`; then they’d be saved to `stderrgpudetect.txt` and we might get a better picture of how far it gets before hanging.

Also, in case anybody’s wondering: `GPU detection failed: process was terminated by signal` is bogus; clearly `114` is not a valid signal number. That’s just the random garbage returned by `waitpid()` after it failed with `EINTR` when the client main process got killed.

**Comment by AenBleidd** [comment]
Do you have these packages installed?

- boinc-client-nvidia-cuda
- boinc-client-opencl
- mesa-opencl-icd
- ocl-icd-libopencl1

?

Пт, 7 апр. 2023 г. в 19:52, Eli T. Drumm ***@***.***>:

> OK, it's just hanging then. Both running as boinc user and as root user. I
> never get that coproc_info.xml file that @davidpanderson
> <https://github.com/davidpanderson> was looking for.
>
> I have added the boinc user to the video group, but maybe it's a
> permissions issue separate from that? Not sure.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BOINC/boinc/issues/5183#issuecomment-1500504752>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYVTIIERLUN6D6KGT75EBDXABH5TANCNFSM6AAAAAAWV7TISM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Best regards,
Vitalii Koshura

Sent via iPhone

**Comment by davidpanderson** [comment]
boinc --detect_gpus shouldn't hang even if expected libraries are missing.

I'll implement Brian's idea of writing warnings to stderr; that may shed some light.
Otherwise maybe Eli can build a debug version from source and use gdb
(we can help with this if needed)

**Comment by BrianNixon** [comment]
> warnings

And/or some tracepoints to track progress through the code. It’s not as though there’s any need to keep the noise down in that file…

**Comment by davidpanderson** [comment]
I added the stderr writes in a 'dpa_gpu_detect' branch.
We can add more as needed.

**Comment by davidpanderson** [comment]
BTW it should take << 1 sec to complete

**Comment by etdr** [comment]
@AenBleidd  So I'm on Arch, which is probably not officially supported, but I have the equivalents of those libraries installed according to the Arch Wiki.

@davidpanderson I can try to clone that branch, build, and report back

**Comment by etdr** [comment]
OK, so with that enabled (plus a few "test" fprintfs from me) here's what I get:

![image](https://user-images.githubusercontent.com/3358909/230749677-d0f36a77-5676-44e0-83c4-4a66c8786403.png)

So I guess the error is happening sometime after the `libaticalrt.so` isn't found, yes? Do I need this file using the current version of BOINC? I would rather use OpenCL, but I probably am misunderstanding something.

**Update:** it's something in the opencl detection. I'll keep digging.

**Update:** It's a problem with this line of code:
```cpp
ciErrNum = (*p_clGetPlatformIDs)(MAX_OPENCL_PLATFORMS, platforms, &num_platforms);
```
Does this mean I need to get in touch with the OpenCL people? ChatGPT tells me this is calling the `clGetPlatformIds` function from OpenCL.

**Comment by etdr** [comment, marked as answer]
OK well by following the instructions in [this thread](https://forum.manjaro.org/t/opencl-code-hangs-on-clgetplatformids/137224) (uninstalling some intel stuff) I was able to get it to work. I guess the problem is with OneAPI. Sorry about all this, but thank you for the help!

**Comment by AenBleidd** [comment]
@etdr, thank you for testing this and sharing the solution.
Since this is not an issue on BOINC side, I'm converting this issue to discussion.

**Comment by BrianNixon** [comment]
> this is not an issue on BOINC side

The root cause might lie elsewhere, but to the user it looks like a bug in BOINC. The client _could_ handle this situation better (with both improved diagnostics and a timeout in waiting for the detection to complete).

**Comment by davidpanderson** [comment]
Probably the easiest thing would be that if the GPU detection process hasn't exited in 10 seconds,
the client considers it failed and goes on.  I'll do this.

**Reply by GreenReaper** [reply]
I think this might have caused a GPU Missing status blocking GPU tasks when starting on a system running an SSD diagnostic involving heavy read load. This is kind of an unusual situation but I guess it might happen more frequently on HDD. Restarting the client fixed it. Just noting it here in case anyone stumbles across it.
